test/Dapr.E2E.Test/DaprApp.cs

```diff
@@ -0,0 +1,99 @@
+﻿// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+using System;
+using System.Net;
+using System.Net.Sockets;
+using System.Linq;
+using System.Runtime.InteropServices;
+
+namespace Dapr.E2E.Test
+{
+    public class DaprApp
+    {
+        // private string pathSeparator;
+        static string shellExeName = SetShell();
+        private string appId;
+        private int appPort;
+        const string outputToMatchOnStart = "You're up and running";
+        const string outputToMatchOnStop = "app stopped successfully";
+
+        static char pathSeparator = System.IO.Path.DirectorySeparatorChar;
+
+        private static string SetShell()
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return "cmd.exe";
+            }
+            return "/bin/bash";
+        }
+
+        public DaprApp(string appId, int appPort = 0)
+        {
+            this.appId = appId;
+            this.appPort = appPort;
+        }
+
+        public (string, string) Start()
+        {
+            var (httpPort, grpcPort, metricsPort) = GetFreePorts();
+            var componentsPath = string.Format($"..{DaprApp.pathSeparator}..{DaprApp.pathSeparator}..{DaprApp.pathSeparator}..{DaprApp.pathSeparator}..{DaprApp.pathSeparator}test{DaprApp.pathSeparator}Dapr.E2E.Test{DaprApp.pathSeparator}components");
+            var daprStartCommand = string.Format($"dapr run --app-id {appId} --dapr-http-port {httpPort} --dapr-grpc-port {grpcPort} --metrics-port {metricsPort} --components-path .{DaprApp.pathSeparator}{componentsPath}");
+            if(this.appPort != 0)
+            {
+                daprStartCommand += string.Format($" --app-port {appPort}");
+            }
+            var projectPath = string.Format($"..{DaprApp.pathSeparator}..{DaprApp.pathSeparator}..{DaprApp.pathSeparator}..{DaprApp.pathSeparator}..{DaprApp.pathSeparator}test{DaprApp.pathSeparator}Dapr.E2E.Test.App{DaprApp.pathSeparator}Dapr.E2E.Test.App.csproj");
+            daprStartCommand += string.Format($" -- dotnet run --project {projectPath}");
+            var daprStart = new ShellCommand()
+            {
+                ShellExeName = DaprApp.shellExeName,
+                Command = daprStartCommand,
+                OutputToMatch = outputToMatchOnStart,
+                Timeout = 5000
+            };
+            daprStart.Run();
```

Comment:
I'd suggest using `System.Diagnostics.Process` unless you specifically need shell functionality. https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process?view=net-5.0

Reply:
Or: https://github.com/dotnet/command-line-api/blob/main/src/System.CommandLine/Invocation/Process.cs

This is a little more full-featured and we could just add a dependency on `System.CommandLine`.